### PR TITLE
test: Check that a block rejected due to too-new time is later accepted in p2p_invalid_block

### PR DESCRIPTION
**Pull request by luke-jr**
I was unsure if the code sufficiently avoided caching a time-too-new rejection, so wrote this test to check it.

It looks like despite only exempting `BLOCK_MUTATED`, it is still okay because header failures never cache block invalidity.

This test will help ensure that if this ever changes, `BLOCK_TIME_FUTURE` gets excluded at the same time.

Note: Branch is a clean merge to 0.19 and master

### CODE REVIEW

File: test/functional/p2p_invalid_block.py [hunk 1]
```diff
@@ -133,5 +134,20 @@ def run_test(self):
...
+        self.log.info("Test accepting identical block after rejecting it due to a future timestamp.")
+
+        t = int(time.time())
+        node.setmocktime(t)
+        block = create_block(tip, create_coinbase(height), t + 7201)
+        block.hashMerkleRoot = block.calc_merkle_root()
+        block.solve()
+        # NOTE: Need force_send because the block will get rejected without a getdata otherwise
+        node.p2p.send_blocks_and_test([block], node, force_send=True, success=False, reject_reason='time-too-new')#, expect_disconnect=True)
+        #self.reconnect_p2p()
```
Comment (jimmysong) [inline test/functional/p2p_invalid_block.py]: take out comment?

**Review by jimmysong — COMMENTED** [review]
LGTM

File: test/functional/p2p_invalid_block.py [hunk 2]
```diff
@@ -11,6 +11,7 @@
 re-requested.
```
Comment (gzhao408) [inline test/functional/p2p_invalid_block.py:11]: Should update description to include the added test?

File: test/functional/p2p_invalid_block.py [hunk 3]
```diff
@@ -133,5 +134,20 @@ def run_test(self):
...
+
+        self.log.info("Test accepting identical block after rejecting it due to a future timestamp.")
+
+        t = int(time.time())
+        node.setmocktime(t)
+        block = create_block(tip, create_coinbase(height), t + 7201)
+        block.hashMerkleRoot = block.calc_merkle_root()
+        block.solve()
+        # NOTE: Need force_send because the block will get rejected without a getdata otherwise
+        node.p2p.send_blocks_and_test([block], node, force_send=True, success=False, reject_reason='time-too-new')#, expect_disconnect=True)
```
Comment (gzhao408) [inline test/functional/p2p_invalid_block.py:146]: Stray in-line comment, and I'm pretty sure we don't expect disconnect?

**Review by glozow — COMMENTED** [review]
ACK with nits, mostly comment-related. It seems like `feature_block.py` and `mining_basic.py` both test too-new blocks but don't check that they're accepted later. I'm not sure what the difference is between `invalid_block.py` and `feature_block.py`; what's the reasoning for putting it here?

**Comment by fanquake** [comment]
Two concept ACKs, but this needs review comments addressing, added "waiting for author".

**Comment by fanquake** [comment]
Basically two months since the last ping and "waiting for author" addition. Going to close this as "up for grabs" if anyone wants to pick it up.

**Comment by MarcoFalke** [comment]
picked up in #22120

**Pull request closed without merging**